README.md

```diff
@@ -0,0 +1,143 @@
+# Collections for PHP with focus on Immutability and Functional Programming
+[![Build Status](https://github.com/bonami/collections/workflows/CI/badge.svg)
+[![Latest Stable Version](https://poser.pugx.org/bonami/collections/v/stable)](https://packagist.org/packages/bonami/collections)
+[![License](https://poser.pugx.org/bonami/collections/license)](https://packagist.org/packages/bonami/collections)
+
+**TODO** table of contents
+
+## Motivation
+
+Why yet another collections library for PHP? Native PHP arrays or SPL structures like SplFixedArray or SplObjectStorage(and other) are mutable and has very strange interfaces and behaviors. They often represents more data structures at once (eg. SplObjectStorage represents both Set and Map) and theirs interfaces are designed for classic imperative approach.
+
+We tried to design interfaces of our structures, to be focused on declarative approach leveraging functional programing. For more safety, we designed structures to be immutable (we have some mutables as well, because sometime it is necessary for optimization reasons)
+
+## Show me the code!
+
+Code can express more then thousands of lines, so here are some simple examples:
+
+### Filtering Person DTOs and extracting some information 
+
+```php
+use Bonami\Collection\ArrayList;
+
+class Person {
+
+	private string $name;
+	private int $age;
+
+	public function __construct(string $name, int $age) {
+		$this->name = $name;
+		$this->age = $age;
+	}
+
+	public function getName(): string {
+		return $this->name;
+	}
+
+	public function getAge(): int {
+		return $this->age;
+	}
+
+}
+
+$persons = ArrayList::of(new Person('John', 31), new Person('Jacob', 22), new Person('Arthur', 29));
+$names = $persons
+	->filter(fn (Person $person): bool => $person->getAge() <= 30)
+	->sort(fn (Person $a, Person $b): int => $a->getName() <=> $b->getName())
+	->map(fn (Person $person): string => $person->getName())
+	->join(";");
+
+// $names = "Arthur;Jacob"
+```
+
+
+**TODO** link on more examples into separate repository
+
+### Generating combinations
+
+```php
+use Bonami\Collection\ArrayList;
+
+$colors = ArrayList::fromIterable(['red', 'green', 'blue']);
+$objects = ArrayList::fromIterable(['car', 'pencil']);
+
+$coloredObjects = ArrayList::fromIterable($colors)
+	->flatMap(fn (string $color) => $objects->map(fn (string $object) => "{$color} {$object}"))
+
+// $coloredObjects = ArrayList::of('red car', 'red pencil', 'green car', 'green pencil', 'blue car', 'blue pencil')
+```
+
+### Generating combinations with lift
+
+```php
+use Bonami\Collection\ArrayList;
+
+$concat = fn (string $first, string $second) => "{$first} {$second}";
+$coloredObjects = ArrayList::lift($concat)($colors, $objects);
```

Comment:
Nejsem si jistej, jestli tohle neni trochu kontraproduktivni priklad - abys pochopil, co se tady deje a proc, musis toho imo znat docela dost (applicative, co dela lift, jaka je jeho semantika pro List)
Nechci teda podcenovat devs nepolibeny tolik FP ;-)

Reply:
Thats why there is an example with and without lift. Maybe (when lift documentation is available) we should link it here.